collectors/carbonblack/collector.js

```diff
@@ -0,0 +1,169 @@
+
+    pawsInitCollectionState(event, callback) {
+        const startTs = process.env.paws_collection_start_ts ?
+            process.env.paws_collection_start_ts :
+            moment().toISOString();
+        const endTs = moment(startTs).add(this.pollInterval, 'seconds').toISOString();
+        const apiNames = JSON.parse(process.env.paws_collector_param_string_1);
+        const initialStates = apiNames.map(apiName => {
+            return {
+                apiName,
+                since: startTs,
+                until: endTs,
+                nextPage: null,
+                poll_interval_sec: 1
+            }
+        });
+        return callback(null, initialStates, 1);
+
+    }
+
+    pawsGetRegisterParameters(event, callback) {
+        const regValues = {
+            carbonblackAPINames: process.env.paws_collector_param_string_1,
+            carbonblackOrgKey: process.env.paws_collector_param_string_2
+        };
+
+        callback(null, regValues);
+    }
+
+    pawsGetLogs(state, callback) {
+
+        let collector = this;
+
+        const clientSecret = JSON.parse(collector.secret);
+        if (!clientSecret) {
+            return callback("The Client Secret was not found!");
+        }
+        
+        const clientId = JSON.parse(collector.clientId);
+        if (!clientId) {
+            return callback("The Client ID was not found!");
+        }
+        
+        const apiEndpoint = process.env.paws_endpoint.replace(/^https:\/\/|\/$/g, '');
+        const orgKey = process.env.paws_collector_param_string_2;
+        const apiDetails = utils.getAPIDetails(state, orgKey);
+
+        if (!apiDetails.url) {
+            return callback("The API name was not found!");
+        }
+
+        typeIdPaths = apiDetails.typeIdPaths;
+        tsPaths = apiDetails.tsPaths;
+
+        console.info(`CABL000001 Collecting data for ${state.apiName} from ${state.since} till ${state.until}`);
+
+        utils.getAPILogs(apiDetails, [], apiEndpoint, state, clientSecret, clientId, process.env.paws_max_pages_per_invocation)
+            .then(({ accumulator, nextPage }) => {
+                let newState;
+                if (nextPage === undefined) {
+                    newState = this._getNextCollectionState(state);
+                } else {
+                    newState = this._getNextCollectionStateWithNextPage(state, nextPage);
+                }
+                console.info(`CABL000002 Next collection in ${newState.poll_interval_sec} seconds`);
+                return callback(null, accumulator, newState, newState.poll_interval_sec);
+            })
+            .catch((error) => {
+                return callback(error);
+            });
+
+
+    }
+
+    _getNextCollectionState(curState) {
+        const nowMoment = moment();
+        const curUntilMoment = moment(curState.until);
+
+        // Check if current 'until' is in the future.
+        const nextSinceTs = curUntilMoment.isAfter(nowMoment) ?
```

Comment:
Please convert this to use the common next state progression. like in the o365 collector.

https://github.com/alertlogic/paws-collector/blob/master/collectors/o365/o365_collector.js#L142